src/xref2/link.ml

```diff
@@ -365,6 +365,51 @@ and open_ env parent = function
   | { Odoc_model__Lang.Open.doc; _ } as open_ ->
       { open_ with doc = comment_docs env parent doc }
 
+let warn_on_hidden_representation (id : Id.Type.t)
+    (r : Lang.TypeDecl.Representation.t) =
+  let open Paths.Identifier in
+  let rec read_typ_exp typ_expr =
+    let open Lang.TypeExpr in
+    let open Paths.Path in
+    match typ_expr with
+    | Constr (p, ts) ->
+        is_hidden (p :> Paths.Path.t)
+        || List.exists (fun t -> read_typ_exp t) ts
+    | Poly (_, t) | Alias (t, _) -> read_typ_exp t
+    | Arrow (_, t, t2) -> read_typ_exp t || read_typ_exp t2
+    | Tuple ts | Class (_, ts) -> List.exists (fun t -> read_typ_exp t) ts
+    | _ -> false
+  in
+
+  let internal_cstr_arg t =
+    let open Lang.TypeDecl.Constructor in
+    let open Lang.TypeDecl.Field in
+    match t.args with
+    | Tuple type_exprs ->
+        List.exists (fun type_expr -> read_typ_exp type_expr) type_exprs
+    | Record fields ->
+        List.exists (fun field -> read_typ_exp field.type_) fields
+  in
+
+  let internal_field t =
+    let open Lang.TypeDecl.Field in
+    read_typ_exp t.type_
+  in
+
+  let fmt_cfg = Component.Fmt.{ default with short_paths = true } in
+  match r with
+  | Variant constructors ->
+      if List.exists internal_cstr_arg constructors then
+        Lookup_failures.report_warning "@[<2>Hidden constructors in type '%a'@]"
```

Comment:
I'm not convinced a warning is useful to users. Except maybe, if there's a way to disable it.

This can happen in legit code and I don't think I would change my code based on this warning.
I suggest removing the warning.

Reply:
The problem usually comes up because someone has overridden the default wrapping that dune does, e.g. writing a manual foo.ml(i) for wrapped library foo. So I don't think it happens unless someone is _trying_ to make the interface shaped differently, and in that case we really do want people to fix it, no? Or is there another way this can happen? For an example of the problem I'm trying to fix 'in the wild', have a look at my PR to omd.

Reply:
I don't think Odoc is the best tool to report this problem. Also, if the interface works, then Odoc should render it nicely.

It seems to me that the warning will be annoying 100% of the time in exchange for being useful rarely.

Reply:
I don't understand, what other tool would report this?

Reply:
Odoc definitely isn't the best tool to report this problem, but unfortunately it's not clear what else can. It's very easy for a type to be used without using its name at all. And so the interface might "work" but there's _no_ nice way to render it. 

How common is it to export a constructor but not its argument type? On purpose, that is?

Reply:
We don't have a tool to report this but it has been suggested that the OCaml community needs a linter.
My point is that we don't report this problem at all.

Reply:
The problem is that in the case of omd a suitable fix involved canonical tags, so whatever tool it is that's reporting the problem, it'll have to understand them. Once we've got such a tool I'll happily delete this code.

Reply:
Works for me. We can continue discussing this if it becomes an unsolvable error.